client/sites-dashboard/components/sites-dashboard.tsx

```diff
@@ -131,8 +131,8 @@ export function SitesDashboard( { queryParams: { search, status = 'all' } }: Sit
 							) ) }
 						</SelectDropdown>
 					</FilterBar>
-					{ filteredSites.length > 0 ? (
-						<SitesTable sites={ filteredSites } />
+					{ filteredSites.length > 0 || isLoading ? (
+						<SitesTable isLoading={ isLoading } sites={ filteredSites } />
```

Comment:
I don't really like having to pass the `isLoading` prop to the `SitesTable` component. It feels like we're delegating that decision that affects the behavior of the whole page to an inner component. Here's how I would do it:

1. I would change `SitesTable` to accept a `children` prop, which would be the content of the table, like `<tbody>{children}</tbody>`;
2. In this `SiteDashboard` component, I would modify the expression to something like:

```tsx
{ filteredSites.length > 0 ? <SitesTable>{ sites.map(...) }</SitesTable> :
isLoading ? <SitesTable><SitesTablePlaceholder /></SitesTable> :
<NoSitesMessage {...} /> }
```

3. `<SitesTablePlaceholder />` would be created and would hold:

```tsx
Array( N_LOADING_ROWS )
						.fill( null )
						.map( ( _, i ) => (
							<SitesTableRowLoading
								columns={ 5 }
								delayMS={ i * 150 }
								logoProps={ { width: 108, height: 78 } }
							/>
						) )
```

Let me know your thoughts on this approach. Another possibility would be to remove the conditionals that prevent the rendering of `<SitesTable />` and then making the decision to display it or not inside the component itself. That's more obscure to me and what's happening is not immediately visible, so I don't like this approach that much...

Reply:
> It feels like we're delegating the decision that affects the behavior of the whole page to an inner component.

The `isLoading` prop seems good to me. The parent component keeps ownership of the _decision_ about whether to render the loading state, but it's delegating the visual details to the table component. For instance what if the table headings also needed a loading state? (e.g. maybe the number of columns depends on the site data) We'd have to refactor `<SitesDashboard>`, having it switch between `<SitesTableHeadings>` and `<SitesTableHeadingsPlaceholder>`.

I do agree that using flags to switch between modes multiple modes for a component can get out of hand! Some of the components in the signup code have lots of modes and it is pretty confusing. But I think a flag for "render a loading version of yourself" can be generally understood in Calypso because it's a design pattern we use throughout the app.